### ISSUE TITLE
R_RISCV_CALL vs R_RISCV_CALL_PLT

### ISSUE THREAD

**Issue by MaskRay** [riscv-non-isa/riscv-elf-psabi-doc#98]
```c
int foo();
int g() { return foo()+1; }
```

This compiles to `R_RISCV_CALL` (`call foo)` in `-fno-pic` mode and `R_RISCV_CALL_PLT` (`call foo@plt`) in -fpie/-fpic mode.

IMHO the distinction is not really useful. We can avoid `R_RISCV_CALL` and use `R_RISCV_CALL_PLT` everywhere (also in `-fno-pic` mode). If the target symbol is non-preemptable (local/hidden/not -shared/etc), the linker can omit PLT creation for `R_RISCV_CALL_PLT`. AFAICT, the only differences are these lines:

```c
// binutils-gdb/bfd/elfnn-riscv.c
	case R_RISCV_CALL_PLT:
	  /* This symbol requires a procedure linkage table entry.  We
	     actually build the entry in adjust_dynamic_symbol,
	     because this might be a case of linking PIC code without
	     linking in any dynamic objects, in which case we don't
	     need to generate a procedure linkage table after all.  */

	  if (h != NULL)
	    {
	      h->needs_plt = 1;
	      h->plt.refcount += 1;
	    }
	  break;

	case R_RISCV_CALL:
	  /* Handle a call to an undefined weak function.  This won't be
	     relaxed, so we have to handle it here.  */
	  if (h != NULL && h->root.type == bfd_link_hash_undefweak
	      && h->plt.offset == MINUS_ONE)
	    {
	      /* We can use x0 as the base register.  */
	      bfd_vma insn = bfd_get_32 (input_bfd,
					 contents + rel->r_offset + 4);
	      insn &= ~(OP_MASK_RS1 << OP_SH_RS1);
	      bfd_put_32 (input_bfd, insn, contents + rel->r_offset + 4);
	      /* Set the relocation value so that we get 0 after the pc
		 relative adjustment.  */
	      relocation = sec_addr (input_section) + rel->r_offset;
	    }
```

I don't know much about BFD internals, but it appears the existence of a PLT in -fpie/-fpic code does not make differences. R_RISCV_CALL_PLT is not necessary to force the creation of a PLT if the symbol is non-preemptable.

In -fno-pic code, there may be a so-called "canonical PLT" (created due to an absolute/pc relative relocation to a function). However, its handling has nothing to do with the dichotomy of `R_RISCV_CALL{,_PLT}`.

I don't know what the weak-undef code does.

On x86_64, newer gas/llvm-mc (after r358652) produces `R_X86_64_PLT32` for `call foo` and `call foo@plt`. `R_X86_64_PLT32` can be optimized to `R_X86_64_PC32` if the symbol is non-preemptable.

powerpc32 ELFv1 specifies `R_PPC_LOCAL24PC` `R_PPC_REL24` and `R_PPC_PLTREL24`. powerpc64 ELFv2 just uses `R_PPC64_REL24`.

**Comment by aswaterman** [comment]
CALL is important for performance, as it supports bypassing the PLT when
the call is known to be local to the shared object.

**Comment by MaskRay** [comment]
@aswaterman Such optimizations can be applied on `R_RISCV_CALL_PLT`, too (i386 x86-64 ppc{32,64} have similar optimizations). The preemptability is the property of the symbol, not the property of the relocation type. The evolution from powerpc ELFv1 to ELFv2 told me people did intend to assign such semantics to relocation types, but they found the semantics may be better handled on the symbol side.

**Comment by jrtc27** [comment]
`call foo` always gives `R_RISCV_CALL`; if you want `R_RISCV_CALL_PLT` you have to write `call foo@plt`.

**Comment by richfelker** [comment]
I too see no justification for the distinction. It makes sense only on targets where there's a special ABI for calling into the PLT (typically, a register that must contain the GOT address, e.g. ebx on i386), in which case call points that aren't explicitly calling foo@PLT cannot be directed into a PLT thunk (a proper one - not an issue for the canonical non-PIC PLT thunk in a non-PIE main executable), and must be a link error (or a textrel) if the symbol is not locally defined.

**Comment by MaskRay** [comment]
@jrtc27 Sorry I made a mistake. I've changed the first sentence to use a C example instead.

**Comment by jrtc27** [comment]
> I too see no justification for the distinction. It makes sense only on targets where there's a special ABI for calling into the PLT (typically, a register that must contain the GOT address, e.g. ebx on i386), in which case call points that aren't explicitly calling foo@PLT cannot be directed into a PLT thunk (a proper one - not an issue for the canonical non-PIC PLT thunk in a non-PIE main executable), and must be a link error (or a textrel) if the symbol is not locally defined.

Sometimes it's useful to ensure this specific call cannot be pre-empted. The FreeBSD libc at least makes use of this; you can work around the issue by defining a local symbol alias, but not distinguishing between `call foo` and `call foo@plt` would break this.

**Comment by aswaterman** [comment]
The RISC-V glibc port relies on the same property @jrtc27 mentioned.

**Comment by jrtc27** [comment]
Also I suppose a reason for the compiler *not* to emit `call foo@plt` with `-fno-PIC` is that it may have performed optimisations on the assumption that `foo` cannot be preempted, which is no longer true if you link it as a shared library (and that may well link just fine, especially if you used -fPIE and aren't using global non-static variables).

**Comment by jim-wilson** [comment]
The undef weak check is to prevent relocation truncated errors when you call a weak function that does not get defined, and your code is compiled for an address above 0x80000000 on a 64-bit target.  For an undefined weak function, we need to make the target address 0.  But this is a pc-relative reloc, which has a 2GB range (medlow/medany) which means we can't actually reach 0 from any address above 0x80000000.  So we detect this case and rewrite the instruction and reloc to be x0 relative instead of pc-relative, and then we can resolve the reloc without error as address 0 is reachable from x0.  There are testcases for this in the ld testsuite, see for instance ld/testsuite/ld-riscv-elf/weakref*.

As a practical matter, since all linux code is compiled for a base address of 0, this only get triggered by embedded developers, so we only needed a fix for R_RISCV_CALL.

**Comment by jim-wilson** [comment]
We can't drop support for R_RISCV_CALL as the ABI was fixed when we upstreamed glibc.

We could perhaps prefer R_RISCV_CALL_PLT and deprecate R_RISCV_CALL in future toolchains.  However, this isn't a bug fix, this is an enhancement.  As a practical matter, we only have one person supporting the entire toolchain, so there is no time available to do enhancement work like this.  There is only time available to fix stuff that is broken.  Perhaps at some point in the future we will have people available to look at issues like this.
